src/Nethermind/Nethermind.Blockchain/Bloom/BloomStorage.cs

```diff
@@ -0,0 +1,374 @@
+                    var levelBloom = levelBlooms[index];
+                    levelBloom.Level.Store(lastBlockNumber, levelBloom.Bloom);
+                }
+                
+                MigratedBlockNumber += i;
+                Set(MigrationBlockNumberKey, MigratedBlockNumber);
+            }
+            
+            if (MigratedBlockNumber >= MinBlockNumber - 1)
+            {
+                MinBlockNumber = 0;
+                Set(MinBlockNumberKey, MinBlockNumber);
+            }
+        }
+
+        private void Set(Keccak key, long value)
+        {
+            _bloomInfoDb.Set(key, value.ToBigEndianByteArrayWithoutLeadingZeros());
+        }
+
+        public void Dispose()
+        {
+            for (int i = 0; i < _storageLevels.Length; i++)
+            {
+                _storageLevels[i].Dispose();
+            }
+        }
+
+        public IBloomEnumeration GetBlooms(long fromBlock, long toBlock) => new BloomEnumeration(_storageLevels, Math.Max(fromBlock, MinBlockNumber), Math.Min(toBlock, MaxBlockNumber));
+
+        private class BloomStorageLevel : IDisposable
+        {
+            public readonly byte Level;
+            public readonly int LevelElementSize;
+            public readonly Average Average = new Average();
+            
+            private readonly IFileStore _fileStore;
+            private readonly bool _migrationStatistics;
+            private readonly LruCache<long, Core.Bloom> _cache;
+            private readonly byte[] _bytes = new byte[Core.Bloom.ByteLength];
+
+            public BloomStorageLevel(IFileStore fileStore, in byte level, in int levelElementSize, in int levelMultiplier, bool migrationStatistics)
+            {
+                _fileStore = fileStore;
+                Level = level;
+                LevelElementSize = levelElementSize;
+                _migrationStatistics = migrationStatistics;
+                _cache = new LruCache<long, Core.Bloom>(levelMultiplier);
+            }
+
+            public void Store(long blockNumber, Core.Bloom bloom)
+            {
+                long bucket = GetBucket(blockNumber);
+                
+                var existingBloom = _cache.Get(bucket);
+                if (existingBloom == null)
+                {
+                    var bytesRead = _fileStore.Read(bucket, _bytes);
+                    var bloomRead = bytesRead == Core.Bloom.ByteLength;
+                    existingBloom = bloomRead ? new Core.Bloom(_bytes) : new Core.Bloom();
+                }
+
+                existingBloom.Accrue(bloom);
+
+                _fileStore.Write(bucket, existingBloom.Bytes);
+                _cache.Set(bucket, existingBloom);
+            }
+            
+
+            private static uint CountBits(Core.Bloom bloom) => (uint) bloom.Bytes.AsSpan().CountBits();
+
+            public long GetBucket(long blockNumber) => blockNumber / LevelElementSize;
+
+            public IFileReader GetReader() => _fileStore.GetFileReader();
```

Comment:
strange to see it public (FIleReader)

Reply:
public in a private class - so like internal